File: packages/react-native/android/src/main/java/expo/modules/motionprobe/MotionProbeModule.kt
package expo.modules.motionprobe

import android.os.Build
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class StartOptions : Record {
  @Field
  val maxDurationMs: Double = 15000.0

  @Field
  val resolveEveryFrames: Int = 6

  /** Sample grid size for occlusion (n × n points); 0 disables occlusion detection. */
  @Field
  val occlusionGrid: Int = 6
}

class MotionProbeModule : Module() {
  private val recorder = MotionRecorder()

  override fun definition() = ModuleDefinition {
    Name("MotionProbe")

    AsyncFunction("getInfo") {
      return@AsyncFunction mapOf(
        "platform" to "android",
        "columns" to MotionRecorder.COLUMNS,
        "osVersion" to Build.VERSION.RELEASE,
        "deviceName" to Build.MODEL,
      )
    }

    AsyncFunction("start") { targets: List<String>, options: StartOptions ->
      return@AsyncFunction recorder.start(
        appContext.currentActivity,
        targets,
        options.maxDurationMs,
        options.resolveEveryFrames,
        options.occlusionGrid,
      )
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("drain") {
      return@AsyncFunction recorder.drain()
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("stop") { reason: String ->
      return@AsyncFunction recorder.stop(reason)
    }.runOnQueue(Queues.MAIN)

    OnDestroy {
      recorder.stop("cancelled")
    }
  }
}

File: packages/react-native/ios/MotionProbeModule.swift
import ExpoModulesCore
import UIKit

struct StartOptions: Record {
  @Field var maxDurationMs: Double = 15000
  @Field var resolveEveryFrames: Int = 6
  /// Sample grid size for occlusion (n × n points); 0 disables occlusion detection.
  @Field var occlusionGrid: Int = 6
}

public final class MotionProbeModule: Module {
  private let recorder = MotionRecorder()

  public func definition() -> ModuleDefinition {
    Name("MotionProbe")

    AsyncFunction("getInfo") { () -> [String: Any] in
      return [
        "platform": "ios",
        "columns": MotionRecorder.columns,
        "osVersion": UIDevice.current.systemVersion,
        "deviceName": UIDevice.current.name,
        "maximumFramesPerSecond": UIScreen.main.maximumFramesPerSecond,
      ]
    }.runOnQueue(.main)

    AsyncFunction("start") { (targets: [String], options: StartOptions) -> [String: Any] in
      return self.recorder.start(
        targets: targets,
        maxDurationMs: options.maxDurationMs,
        resolveEveryFrames: options.resolveEveryFrames,
        occlusionGrid: options.occlusionGrid
      )
    }.runOnQueue(.main)

    AsyncFunction("drain") { () -> [String: Any] in
      return self.recorder.drain()
    }.runOnQueue(.main)

    AsyncFunction("stop") { (reason: String) -> [String: Any] in
      return self.recorder.stop(reason: reason)
    }.runOnQueue(.main)

    OnDestroy {
      DispatchQueue.main.async { [recorder] in
        _ = recorder.stop(reason: "cancelled")
      }
    }
  }
}
